src/lib/io/sol-uart-impl-linux.c

```diff
@@ -172,72 +200,90 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
-        sol_fd_del(uart->async.rx_fd_handler);
-    if (uart->async.tx_fd_handler)
-        sol_fd_del(uart->async.tx_fd_handler);
+
+    if (uart->fd_handler)
+        sol_fd_del(uart->fd_handler);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&uart->pending_blobs, blob, i)
+        sol_blob_unref(blob);
+    sol_ptr_vector_clear(&uart->pending_blobs);
     close(uart->fd);
     free(uart);
 }
 
-static void
-uart_tx_dispatch(struct sol_uart *uart, int status)
+SOL_API int
+sol_uart_write(struct sol_uart *uart, struct sol_blob *blob)
 {
-    uart->async.tx_fd_handler = NULL;
-    if (!uart->async.tx_cb)
-        return;
-    uart->async.tx_cb((void *)uart->async.tx_user_data, uart,
-        (uint8_t *)uart->async.tx_buffer, status);
-}
+    int r;
+    bool fd_created = false;
 
-static bool
-uart_tx_callback(void *data, int fd, uint32_t active_flags)
-{
-    struct sol_uart *uart = data;
-    int ret;
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(blob, -EINVAL);
 
-    if (active_flags & FD_ERROR_FLAGS) {
-        SOL_ERR("Error flag was set on UART file descriptor %d.", fd);
-        ret = -1;
-        goto error;
-    }
+    if (!uart->fd_handler) {
+        uart->fd_handler = sol_fd_add(uart->fd,
+            FD_ERROR_FLAGS | SOL_FD_FLAGS_OUT, uart_fd_handler_callback, uart);
+        SOL_NULL_CHECK(uart->fd_handler, -ENOMEM);
+        fd_created = true;
+    } else {
+        bool err;
 
-    if (uart->async.tx_index == uart->async.tx_length) {
-        uart_tx_dispatch(uart, uart->async.tx_index);
-        return false;
+        err = sol_fd_add_flags(uart->fd_handler, SOL_FD_FLAGS_OUT);
+        SOL_EXP_CHECK(!err, -EINVAL);
     }
 
-    ret = write(fd, uart->async.tx_buffer + uart->async.tx_index,
-        uart->async.tx_length - uart->async.tx_index);
-    if (ret < 0) {
-        SOL_ERR("Error when writing to file descriptor %d.", fd);
-        goto error;
+    r = -EOVERFLOW;
+    SOL_NULL_CHECK_GOTO(sol_blob_ref(blob), err_ref);
+
+    r = sol_ptr_vector_append(&uart->pending_blobs, blob);
+    SOL_INT_CHECK_GOTO(r, < 0, err_append);
+
+    return 0;
+
+err_append:
+    sol_blob_unref(blob);
+err_ref:
+    if (fd_created) {
+        sol_fd_del(uart->fd_handler);
+        uart->fd_handler = NULL;
     }
+    return r;
+}
 
-    uart->async.tx_index += ret;
-    return true;
+SOL_API int
+sol_uart_read(struct sol_uart *uart, struct sol_buffer *buf)
+{
+    int r, total;
+
+    SOL_NULL_CHECK(uart, -EINVAL);
+    SOL_NULL_CHECK(buf, -EINVAL);
 
-error:
-    uart_tx_dispatch(uart, ret);
-    return false;
+    r = ioctl(uart->fd, FIONREAD, &total);
+    SOL_INT_CHECK(r, < 0, r);
+    return (int)sol_util_fill_buffer(uart->fd, buf, total);
 }
 
-SOL_API bool
-sol_uart_write(struct sol_uart *uart, const uint8_t *tx, unsigned int length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data)
+SOL_API int
+sol_uart_get_pending_bytes(struct sol_uart *uart, size_t *pending_bytes)
```

Comment:
at a second look, this function should reflect it's write bytes, not read bytes. So rename to `sol_uart_get_pending_write_bytes()`